anomalib/models/ganomaly/torch_model.py

```diff
@@ -0,0 +1,254 @@
+
+
+import math
+
+from torch import Tensor, nn
+
+
+class Encoder(nn.Module):
+    """Encoder Network.
+
+    Args:
+        input_size (int): Size of input image
+        latent_vec_size (int): Size of latent vector z
+        num_input_channels (int): Number of input channels in the image
+        n_features (int): Number of features per convolution layer
+        extra_layers (int): Number of extra layers since the network uses only a single encoder layer by default.
+            Defaults to 0.
+    """
+
+    def __init__(
+        self,
+        input_size: int,
+        latent_vec_size: int,
+        num_input_channels: int,
+        n_features: int,
+        extra_layers: int = 0,
+        add_final_conv_layer: bool = True,
+    ):
+        super().__init__()
+
+        assert input_size % 16 == 0, "Input size should be a multiple of 16"  # Why?
+
+        self.input_layers = nn.Sequential()
+
+        self.input_layers.add_module(
+            f"initial-conv-{num_input_channels}-{n_features}",
+            nn.Conv2d(num_input_channels, n_features, kernel_size=4, stride=2, padding=1, bias=False),
+        )
+        self.input_layers.add_module(f"initial-relu-{n_features}", nn.LeakyReLU(0.2, inplace=True))
+
+        # Extra Layers
+        self.extra_layers = nn.Sequential()
+
+        for layer in range(extra_layers):
+            self.extra_layers.add_module(
+                f"extra-layers-{layer}-{n_features}-conv",
+                nn.Conv2d(n_features, n_features, kernel_size=3, stride=1, padding=1, bias=False),
+            )
+            self.extra_layers.add_module(f"extra-layers-{layer}-{n_features}-batchnorm", nn.BatchNorm2d(n_features))
+            self.extra_layers.add_module(f"extra-layers-{layer}-{n_features}-relu", nn.LeakyReLU(0.2, inplace=True))
+
+        # Create pyramid features to reach latent vector
+        self.pyramid_features = nn.Sequential()
+        input_size = input_size // 2
+        while input_size > 4:
+            in_features = n_features
+            out_features = n_features * 2
+            self.pyramid_features.add_module(
+                f"pyramid-{in_features}-{out_features}-conv",
+                nn.Conv2d(in_features, out_features, kernel_size=4, stride=2, padding=1, bias=False),
+            )
+            self.pyramid_features.add_module(f"pyramid-{out_features}-batchnorm", nn.BatchNorm2d(out_features))
+            self.pyramid_features.add_module(f"pyramid-{out_features}-relu", nn.LeakyReLU(0.2, inplace=True))
+            n_features = out_features
+            input_size = input_size // 2
+
+        # Final conv
+        if add_final_conv_layer:
+            self.final_conv_layer = nn.Conv2d(
+                n_features, latent_vec_size, kernel_size=4, stride=1, padding=0, bias=False
+            )
+
+    def forward(self, input_tensor: Tensor):
+        """Return latent vectors."""
+
+        output = self.input_layers(input_tensor)
+        if len(self.extra_layers) > 0:
```

Comment:
I don't think this check is needed. If the sequential module is empty, calling it will just return the unaltered input.

Reply:
Thanks! I didn't know this. I have updated the PR.